lib/charms/operator_libs_linux/v0/grub.py

```diff
@@ -29,4 +37,259 @@
+    """Exception representing value validation error."""
+
+    def __init__(self, key: str, message: str) -> None:
+        self._key = key
+        self._message = message
+        super().__init__(message)
+
+    @property
+    def key(self) -> str:
+        """Get name of key, which cause this error."""
+        return self._key
+
+    @property
+    def message(self) -> str:
+        """Get error message."""
+        return self._message
+
+
+class IsContainerError(Exception):
+    """Exception if local machine is container."""
+
+
+class ApplyError(Exception):
+    """Exception if applying new config failed."""
+
+
+def _parse_config(data: str) -> Dict[str, str]:
+    """Parse config file lines.
+
+    This function is capable to update single key value if it's used with $ symbol.
+    """
+    config = {}
+    lines = data.splitlines()
+    for line in lines:
+        line = line.strip()
+        if not line or line.startswith("#"):
+            logger.debug("skipping line `%s`", line)
+            continue
+
+        key, value = line.split("=", 1)
+        if key in config:
+            logger.warning("key %s is duplicated in config", key)
+
+        config[key] = value
+
+    return config
+
+
+def _load_config(path: Path) -> Dict[str, str]:
+    """Load config file from /etc/default/grub.d/ directory."""
+    if not path.exists():
+        raise FileNotFoundError("grub config file %s was not found", path)
+
+    with open(path, "r", encoding="UTF-8") as file:
+        data = file.read()
+        config = _parse_config(data)
+
+    logger.info("grub config file %s was loaded", path)
+    logger.debug("config file %s", config)
+    return config
+
+
+def _save_config(path: Path, config: Dict[str, str], header: str = CONFIG_HEADER) -> None:
+    """Save grub config file."""
+    if path.exists():
+        logger.debug("grub config %s already exist and it will overwritten", path)
+
+    context = [f"{key}={value}" for key, value in config.items()]
+    with open(path, "w", encoding="UTF-8") as file:
+        file.writelines([header, *context])
+
+    logger.info("grub config file %s was saved", path)
+
+
+def is_container() -> bool:
+    """Help function to see if local machine is container."""
+    # TODO: implement it
+    return False
+
+
+class GrubConfig(Mapping[str, str]):
+    """Grub config object.
+
+    This object will load current configuration option for grub and provide option
+    to update it with simple validation.
+
+    There is only one public function `update`, which should handle everything. It will
+    raise exception if there is already different value configured and return set of
+    changed keys to help charm check which keys where changed.
+    """
+
+    _lazy_data: Optional[Dict[str, str]] = None
+
+    def __init__(self, charm_name: Optional[str] = None) -> None:
+        """Initialize the grub config."""
+        self._charm_name = charm_name
+
+    def __contains__(self, key: str) -> bool:
```

Comment:
Would it be simpler to just have `GrubConfig` subclass `dict` directly?

Reply:
That said, I often find these "smart objects" cause more confusion than they bring value. It's kind of a dict, but not quite. Worth considering a dumber/simpler API that deals in plain dicts, perhaps top-level functions like `read_config() -> dict` and `update_config(dict)`?

Reply:
It's not subclass of dict, only due to usage of lazy loading. If you check example in lib docstring on in this PR.
Basically the idea is to define `self.grub = GrubConfig()` in charm `__init__`, so it can be access in any function. Some charm may want to use `apply` in action instead of install, some maybe will do some configuration via config_changed.